### PR TITLE
Activate hairpin mode when sync'ing pods. Fixes #13375

### LINKED ISSUE #13375: Pod-to-self networking require hairpin mode
Linked to #3760, when using the iptables proxy instead of userspace, a pod can connect to a service IP it serves, looping back to itself. In this case, the Linux bridge implementation stops the packet from being processed because it then has the same input and output interfaces. 

The solution to this problem is to enable the hairpin of each port of the virtual bridge where the pods are connected. It should be equivalent to the following shell script:

```
# To get traffic from a pod to itself via a service:
for intf in $(ip link list | grep veth | cut -f2 -d:); do brctl hairpin cbr0 $intf on; done
```

/cc @thockin

### PR DESCRIPTION
**Pull request by evanphx**
Fixes #13375. This gets the ability to use iptables proxying one step closer.

### CODE REVIEW

**Comment by eparis** [comment]
Assigning to @thockin since he is the guy reworking this area. But I can start by saying this can't possible be right, can it? cbr0 doesn't exist on any system using flannel. Doesn't this type of thing need to be handled by the thing implementing the network functions (aka flannel for many non-GCE setups) and not by the kubelet?

**Comment by evanphx** [comment]
Well, kubelet is hardcoded to expect and use cbr0 it looks like. If the usage of cbr0 is gated by configuration, the call to setHairpin can also be gated.

**Comment by evanphx** [comment]
Which providers are using flannel by default? Or is it something users can configure on cluster setup? AWS doesn't.

**Comment by evanphx** [comment]
Ah looks like CoreOS?

**Comment by eparis** [comment]
ok to test

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -2183,6 +2188,28 @@ func (kl *Kubelet) reconcileCBR0(podCIDR string) error {
 	return kl.shaper.ReconcileInterface()
 }
 
+func (kl *Kubelet) setHairpin() {
+	const dir = "/sys/devices/virtual/net/cbr0/brif"
```
Comment (thockin) [inline pkg/kubelet/kubelet.go]: We can avoid hard-coding cbr0 by going broader and setting `/sys/devices/virtual/net/*/brport/hairpin_mode` - the sad truth is that hard-coding cbr0 is not right, and while we do it in some places, we REALLY need to not make it worse.

The risk being that we get some veth interfaces that we don't want, but that seems less bad to me.

**Comment by thockin** [comment]
@MikaelCluseau because overlap

It's not right for flannel, but the revised for should not be harmful.  To be honest, though, I need to fully test flannel and this new proxier.  There may be more lurking.

**Comment by MikaelCluseau** [comment]
To me, getting the right bridge will require backend-specific code.

**Comment by evanphx** [comment]
Is k8s always in control of what the bridge is? Or do we need to do some probing at first to detect it?




On kubelet start, we could start a pause container, see what bridge just got a new interface, then tear down the pause. Then we know exactly which bridge is in use.




Or should we except the container backend to just provide the bridge name and leave it up to however that needs to work. 




I'm happy to flesh it out either way, just want some guidance first. 




Thanks!

**Comment by MikaelCluseau** [comment]
@evanphx The problem is that we overlap with different approaches. I get the host's pair interface of the container's main interface, that is eth0 in docker (see [here](http://docs.docker.com/articles/networking/#container-networking)). My PR is #13628.

**Comment by thockin** [comment]
No, kubernetes is not always in charge of setting up the bridge - this is
why I am advocating not to spell out "cbr0" anywhere.  I;d rather just
blast a 1 into hairpin_mode for all bridge-ports that support it, and
ignore errors.  Before long we will have real plugins and this will be
embedded in there.

**Comment by evanphx** [comment]
I've got no problem closing this in favor of #13628. Just let me know if that's direction to take.

**Comment by thockin** [comment]
We're trying to see if a narrower approach is really feasible - let's give
it a day or two, but then I want to decide and get this in

**Comment by evanphx** [comment]
@thockin Sounds good! Would you like me to revise this PR to remove referencing `cbr0`?

**Comment by thockin** [comment]
If you have time to potentially waste, sure :)

**Comment by eparis** [comment]
Some backends won't use a linux bridge at all.  Just food for thought, not arguing strongly against the blast, but just making this 'yet another requirement on the network implementation' feels like the 'right' way to go....

openshift-sdn, for instance, will let docker create pause and I don't know, it either creates with --net=none or it attaches pause to a linux bridge, but then osdn actually disconnects it entirely from the linux bridge and connects it to an OVS bridge. So setting hairpin on all linux bridges will have no affect what so ever on the containers/pods...

**Comment by MikaelCluseau** [comment]
ummm what does `nsenter -n -t {container-pid} ethtool -S eth0` in this case?

**Comment by thockin** [comment]
I think we should focus on #13628 at this point.

**Comment by thockin** [comment]
Thanks for this PR, though.  Really.

**Comment by evanphx** [comment]
No problem, happy to help. I'm looking to get involved with kubernetes and this seemed like a good place since I'm really interested in get high perf networking out. Is there anything else you need tackled?

**Comment by thockin** [comment]
I'd love to see a better set of tests that were canned and easy to run
which measured kubernetes pod-to-service throughput and latency.  Our GSoC
student @BenTheElder did some tests on the new proxy, but they are not as
repeatable or easy as I would like.  This would give us a chance to compare
the userspace proxy, the iptables proxy, as well as the 3-4 other
implementations people have demo'ed.  It would also let us show off
performance of the various overlay/underlay/etc plugins that are emerging.

If you want something "codier" we can find something there, too, I am sure.

**Pull request closed without merging**